File: src/cpp/modules/tapenade/lstm/lstm.cpp
#include "lstm.h"

// UTILS
// Sigmoid on scalar
double sigmoid(double x)
{
    return 1.0 / (1.0 + exp(-x));
}

// log(sum(exp(x), 2))
double logsumexp(const double* const vect, int sz)
{
    double sum = 0.0;
    int i;

    for (i = 0; i < sz; i++)
    {
        sum += exp(vect[i]);
    }

    sum += 2;
    return log(sum);
}

// LSTM OBJECTIVE
// The LSTM model
void lstm_model(
    int hsize,
    const double* const weight,
    const double* const bias,
    double* hidden,
    double* cell,
    const double* const input
)
{
    double* gates = (double*)malloc(4 * hsize * sizeof(double));
    double* forget = &(gates[0]);
    double* ingate = &(gates[hsize]);
    double* outgate = &(gates[2 * hsize]);
    double* change = &(gates[3 * hsize]);

    int i;
    for (i = 0; i < hsize; i++)
    {
        forget[i] = sigmoid(input[i] * weight[i] + bias[i]);
        ingate[i] = sigmoid(hidden[i] * weight[hsize + i] + bias[hsize + i]);
        outgate[i] = sigmoid(input[i] * weight[2 * hsize + i] + bias[2 * hsize + i]);
        change[i] = tanh(hidden[i] * weight[3 * hsize + i] + bias[3 * hsize + i]);
    }

    for (i = 0; i < hsize; i++)
    {
        cell[i] = cell[i] * forget[i] + ingate[i] * change[i];
    }

    for (i = 0; i < hsize; i++)
    {
        hidden[i] = outgate[i] * tanh(cell[i]);
    }

    free(gates);
}

// Predict LSTM output given an input
void lstm_predict(
    int l,
    int b,
    const double* const w,
    const double* const w2,
    double* s,
    const double* const x,
    double* x2
)
{
    int i;
    for (i = 0; i < b; i++)
    {
        x2[i] = x[i] * w2[i];
    }

    double* xp = x2;
    for (i = 0; i <= 2 * l * b - 1; i += 2 * b)
    {
        lstm_model(b, &(w[i * 4]), &(w[(i + b) * 4]), &(s[i]), &(s[i + b]), xp);
        xp = &(s[i]);
    }

    for (i = 0; i < b; i++)
    {
        x2[i] = xp[i] * w2[b + i] + w2[2 * b + i];
    }
}

// LSTM objective (loss function)
void lstm_objective(
    int l,
    int c,
    int b,
    const double* const main_params,
    const double* const extra_params,
    double* state,
    const double* const sequence,
    double* loss
)
{
    int i, t;
    double total = 0.0;
    int count = 0;
    const double* input = &(sequence[0]);
    double* ypred = (double*)malloc(b * sizeof(double));
    double* ynorm = (double*)malloc(b * sizeof(double));
    const double* ygold;
    double lse;

    for (t = 0; t <= (c - 1) * b - 1; t += b)
    {
        lstm_predict(l, b, main_params, extra_params, state, input, ypred);
        lse = logsumexp(ypred, b);
        for (i = 0; i < b; i++)
        {
            ynorm[i] = ypred[i] - lse;
        }

        ygold = &(sequence[t + b]);
        for (i = 0; i < b; i++)
        {
            total += ygold[i] * ynorm[i];
        }

        count += b;
        input = ygold;
    }

    *loss = -total / count;

    free(ypred);
    free(ynorm);
}

File: src/dotnet/modules/DiffSharpModule/lstm.fs
module lstm

open DiffSharp.AD.Float64
open DotnetRunner.Data
open System.Composition
open DiffSharp.Util

let lstmModel (weight: DV) (bias: DV) (hidden: DV) (cell: DV) (input: DV) : DV * DV =
    let hsize = hidden.Length
    let forget = sigmoid (input .* weight.[0..hsize - 1] + bias.[0..hsize - 1])
    let ingate = sigmoid (hidden .* weight.[hsize..2 * hsize - 1] + bias.[hsize..2 * hsize - 1])
    let outgate = sigmoid (input .* weight.[2 * hsize..3 * hsize - 1] + bias.[2 * hsize..3 * hsize - 1])
    let change = tanh (hidden .* weight.[3 * hsize..4 * hsize - 1] + bias.[3 * hsize..4 * hsize - 1])

    let cell2 = cell .* forget + ingate .* change
    let hidden2 = outgate .* tanh (cell2)
    hidden2, cell2

let lstmPredict (mainParams: DM) (extraParams: DM) (state: DM) (input: DV) : DV * DM =
    let x = input .* extraParams.[0, *]
    let lenState = state.Rows / 2
    let s2list, x2 = [0..lenState - 1]
                        |> List.fold (fun (acc:DV list * DV) i -> 
                            let s, x = acc
                            let h, c = lstmModel mainParams.[2 * i, *] mainParams.[2 * i + 1, *] state.[2 * i, *] state.[2 * i + 1, *] x
                            (c::h::s, h)
                            ) ([], x)
    x2 .* extraParams.[1, *] + extraParams.[2, *], DM.ofRows (List.rev s2list)

let lstmObjective (mainParams: DM) (extraParams: DM) (state: DM) (sequence: DM) : D =
    let lenSeq = sequence.Rows
    let count = sequence.Cols * (sequence.Rows - 1)
    let total = [0..lenSeq - 2]
                |> List.fold (fun (acc:DM * D) i ->
                    let oldState, oldTotal = acc
                    let ypred, newState = lstmPredict mainParams extraParams oldState sequence.[i, *]
                    let ynorm = ypred - log (DV.Sum (exp ypred) + 2.)
                    newState, oldTotal + sequence.[i + 1, *] * ynorm) (state, D 0.)
                |> snd
    -total / count

[<Export(typeof<DotnetRunner.ITest<LSTMInput, LSTMOutput>>)>]
type DiffSharpLSTM() =
    [<DefaultValue>] val mutable input : LSTMInput
    [<DefaultValue>] val mutable packedInput : DV
    [<DefaultValue>] val mutable lstmObjectiveWrapper : DV -> D
    let mutable objective : D = D 0.
    let mutable gradient : DV = DV.empty
     
    interface DotnetRunner.ITest<LSTMInput, LSTMOutput> with
        member this.Prepare(input: LSTMInput) : unit = 
            this.input <- input
            this.packedInput <- Array.map toDV (Array.append input.MainParams input.ExtraParams) |> DV.concat
            let mainParamsSliceCount = 2 * input.LayerCount
            let mainParamsSize = 8 * input.LayerCount * input.CharBits
            let stateDM = input.State |> Seq.ofArray |> Seq.map Seq.ofArray |> toDM
            let sequenceDM = input.Sequence |> Seq.ofArray |> Seq.map Seq.ofArray |> toDM
            this.lstmObjectiveWrapper <- (fun par ->
                let mainParams = DV.splitEqual mainParamsSliceCount par.[0..mainParamsSize - 1] |> DM.ofRows
                let extraParams = DV.splitEqual 3 par.[mainParamsSize..] |> DM.ofRows
                lstmObjective mainParams extraParams stateDM sequenceDM)
            // Let's build DiffSharp internal Reverse AD Trace
            // To do it just calculate the function in the another point
            // Moreover, it forces JIT-compiler to compile the function
            let oldInput = this.packedInput
            this.packedInput <- this.packedInput + 1.
            (this :> DotnetRunner.ITest<LSTMInput, LSTMOutput>).CalculateObjective(1)
            (this :> DotnetRunner.ITest<LSTMInput, LSTMOutput>).CalculateJacobian(1)
            // Put the old input back 
            this.packedInput <- oldInput

        member this.CalculateObjective(times: int) : unit =
            [1..times] |> List.iter (fun _ ->
                objective <- this.lstmObjectiveWrapper this.packedInput
            )

        member this.CalculateJacobian(times: int) : unit =
            [1..times] |> List.iter (fun _ ->
                gradient <- grad this.lstmObjectiveWrapper this.packedInput
            )
            
        member this.Output() : LSTMOutput = 
            let mutable output = new LSTMOutput()
            output.Objective <- convert objective
            output.Gradient <- convert gradient
            output

